Complete elliptic solver initialization with actions

nilsvu opened the pull request:
## Proposed changes

Converts the remaining elliptic initializers into actions (together with #1651). This requires some refactoring of the linear solver that is straight-forward but changes many lines (last commit).

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add compute item for the next linear solver IterationId

nilsvu committed:
Use DG interfaces initializer in elliptic executable

nilsvu committed:
Move action to impose elliptic inhom. b.c. out of InitializeElement

nilsvu committed:
Re-enable test for elliptic inhomogeneous boundary conditions

nilsvu committed:
Fix some namespaces in elliptic code

nilsvu committed:
Make initializing exterior variables optional

These are initialized as a simple tag and updated manually for imposing
boundary conditions. To allow for different ways of imposing boundary
conditions, this commit adds a template parameter flag that makes the
initializer not add this simple tag.

nilsvu commented:
Can I get two reviewers for this? :) Perhaps @moxcodes and one of @nilsdeppe @wthrowe @kidder could you take this one?

This PR finishes the series of PRs that got rid of `InitializeElement` in favor of initialization actions.

moxcodes commented on src/Elliptic/DiscontinuousGalerkin/ImposeInhomogeneousBoundaryConditionsOnSource.hpp:
```diff
@@ -26,7 +26,8 @@ struct Inertial;
 /// \endcond
```
long first line on this commit

moxcodes commented on tests/Unit/Elliptic/DiscontinuousGalerkin/Test_ImposeInhomogeneousBoundaryConditionsOnSource.cpp:
```diff
@@ -78,76 +90,99 @@ struct NumericalFlux {
-
-    const auto box = elliptic::Initialization::BoundaryConditions<
-        Metavariables<1>>::initialize(std::move(arguments_box), runner.cache());
+    ActionTesting::emplace_component_and_initialize<element_array>(
```
with the additional typedefs, this part has every appearance of being identically reproduced for each dimensionality (aside from the comments). Any chance it could be split out to a template function?

nilsvu replied:
good idea 👍

moxcodes commented on src/Elliptic/DiscontinuousGalerkin/ImposeInhomogeneousBoundaryConditionsOnSource.hpp:
```diff
@@ -70,7 +70,7 @@ struct ImposeInhomogeneousBoundaryConditionsOnSource {
   using system = typename Metavariables::system;
```
Does the previous commit actually compile without these namespace fixes? If not, I'd prefer it to be just squashed in so each commit actually compiles.

nilsvu replied:
It does, since the `elliptic::Tags` namespace is not added in this PR yet. It's coming soon though :)

nilsdeppe commented:
Is it not possible to split this into smaller PRs? +1500 -1200 is a huge diff...

nilsvu commented:
@nilsdeppe I‘m not sure I can split the commits any further. But I can make multiple PRs out of them if that helps.

nilsvu committed:
Make linear solver initializer an action

This is the last remaining elliptic initializer to be converted to
actions, so `InitializeElement` can now be deleted. The new linear
solver initializer action can also be used to re-initialize the
linear solver, which allows to perform multiple linear solves in
sequence. This is needed for solving nonlinear equations, where
we need to solve the linearized equations in each step of the
nonlinear solver.

nilsvu committed:
Template linear solvers on the fields they solve

This decouples the linear solvers from the system, since they only need
to know the fields they solve for. It allows using them to solve for any
fields in the DataBox, in particular the linearized fields in each step
of a nonlinear solve.

nilsvu commented:
@nilsdeppe I split the last commit in two, so a third of the changed lines in this PR is now in the very last commit (86eacd2 "Template linear solvers on the fields they solve") which should be trivial to review. I can also move that commit into a separate PR if that helps, but it should be quite uncontroversial. I hope this makes this PR easier to review. All commits should be reasonably small and straight-forward now.

moxcodes commented on src/NumericalAlgorithms/LinearSolver/ConjugateGradient/InitializeElement.hpp:
```diff
@@ -34,55 +35,89 @@ struct InitializeResidual;
+            // iteration ID such that the _next_ iteration ID is zero.
+            std::numeric_limits<size_t>::max(), std::move(residual),
+            db::item_type<LinearSolver::Tags::HasConverged>{}),
+        // Terminate algorithm for now. The reduction will be broadcast to the
```
I'm not certain it is the most intuitive strategy to terminate the algorithm here. Is there a good reason to prefer terminating here over including a separate `TerminatePhase` action in the action list or over using an `is_ready` to determine whether the reduction has been called?

nilsvu replied:
This is how global synchronizations are currently handled in the linear solver, see e.g. `ConjugateGradient/ElementActions.hpp`. I'm happy to discuss alternatives. However, that's a (potential) follow-up PR since it's a more involved change of how the linear solver algorithm works. To your suggestions:
- The approach with the `TerminatePhase` action wouldn't work I think. It's possible that the reduction comes in before the `TerminatePhase` action is reached in the action list, in which case the algorithm wouldn't be restarted.
- `is_ready` would require an additional action, which the user has to remember to add to the action list. I like that this would - in principle - allow doing work while the reduction is processing. But during initialization this hardly matters and during the actual algorithm we need the result of the reduction to proceed with the linear solver step. So I don't currently see the additional complexity for the user compensated by a gain in performance. And going forward the idea is to improve parallelization not by avoiding or optimizing these global synchronization points, but by preconditioning the linear solver in a parallelizable way (i.e. Multigrid with Schwarz smoother).

moxcodes replied:
I see, thanks. I won't request a code change on this, then. My only concern is that it may not be obvious to users of this code creating an initialization action list that the initialization will not proceed after this action until the reduction occurs. If you agree that this is a misunderstanding to avoid, I'll leave it up to you where it's best to make a documentation note -- I'm not sure whether it should be at the start of this detail struct or preceding the typedefs which alias to template specializations of this struct.

nilsvu replied:
Such a note already exists for `prepare_step` in `ConjugateGradient.hpp` and `Gmres.hpp`, so I added it also to `initialize_element`. Please check the fixup :)

moxcodes commented on src/NumericalAlgorithms/LinearSolver/Gmres/InitializeElement.hpp:
```diff
@@ -47,49 +52,93 @@ struct InitializeElement {
+            db::item_type<LinearSolver::Tags::HasConverged>{}),
+        // Terminate algorithm for now. The reduction will be broadcast to the
+        // next action which is responsible for restarting the algorithm.
+        true);
```
same pattern as previous comment about algorithm termination, so if changed above, similar changes should be applied here.

moxcodes commented on tests/Unit/NumericalAlgorithms/LinearSolver/LinearSolverAlgorithmTestHelpers.hpp:
```diff
@@ -109,46 +112,28 @@ struct TestResult {
+                    const Parallel::ConstGlobalCache<Metavariables>& cache,
+                    const int /*array_index*/, const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/) noexcept {
     const auto& A = get<LinearOperator>(cache);
```
It appears you've moved away from the single-character names in the rest of the code. If you're applying that policy throughout, it should be changed here too.

nilsvu replied:
ok :)

moxcodes reviewed (commented):
A few minor questions and comments. Largely, this seems to preserve existing functionality and move around where tags are obtained from, so pretty non-objectionable.

nilsvu commented:
I'd appreciate a second review for this, perhaps @hrueter or @prayush could you review?

moxcodes reviewed (commented):
fixups look good to me, go ahead and squash

hrueter commented on src/NumericalAlgorithms/LinearSolver/ConjugateGradient/InitializeElement.hpp:
```diff
@@ -34,55 +35,89 @@ struct InitializeResidual;
+
+    // Proceed with algorithm
+    Parallel::get_parallel_component<ParallelComponent>(cache)[array_index]
+        .perform_algorithm(true);
```
I think it would be a little bit more clear if this instead of here was executed at the end of the `InitializeResidual`-Action that is calling `InitializeHasConverged`.

nilsvu replied:
That's on another parallel component - the residual monitor singleton. It receives the reduction and broadcasts it back to the element array, which then proceeds with the algorithm.

hrueter reviewed (commented):
The PR looks good so far, but I agree with @moxcodes that the intermediate termination strategy is confusing.

nilsvu commented:
@moxcodes @hrueter fixed a clang-tidy issue so Travis passes. Also fixed an issue in the action testing framework that came up in a test (commit "Mark a mock parallel component function `const`").

nilsdeppe commented on tests/Unit/ActionTesting.hpp:
```diff
@@ -1161,7 +1161,7 @@ class MockProxy {
 
   MockArrayElementProxy<Component, InboxTagList> operator[](
-      const Index& index) {
+      const Index& index) const noexcept {
```
This is incorrect please remove it. 

Edit:

I'm not convinced an overload is right either

nilsvu replied:
My mistake, dropped this commit. Fixed a `const` elsewhere instead.

nilsdeppe commented on tests/Unit/Elliptic/DiscontinuousGalerkin/Test_ImposeInhomogeneousBoundaryConditionsOnSource.cpp:
```diff
@@ -50,13 +57,18 @@ struct System {
   tuples::TaggedTuple<ScalarFieldTag> variables(
       const tnsr::I<DataVector, Dim>& x,
       tmpl::list<ScalarFieldTag> /*meta*/) const noexcept {
+    // Keeping this expression simple to not complicate the computation of the
```
Please make this depend on the the dimensionality.

nilsdeppe commented on src/ParallelAlgorithms/DiscontinuousGalerkin/InitializeInterfaces.hpp:
```diff
@@ -47,6 +47,40 @@ using exterior_compute_tags = tmpl::list<Tags...>;
+template <bool Enable, typename Metavariables>
+struct InitExteriorVarsImpl {
+  template <typename DbTagsList>
+  static auto apply(db::DataBox<DbTagsList>&& box) noexcept {
```
You know the return type here, please write it explicitly since `auto` works in mysterious ways.

nilsdeppe reviewed (changes requested):
I haven't had a chance to look at the last two commits.

nilsvu commented:
Rebased on #1650 and posted fixups

hrueter commented on tests/Unit/Elliptic/DiscontinuousGalerkin/Test_ImposeInhomogeneousBoundaryConditionsOnSource.cpp:
```diff
@@ -205,23 +204,17 @@ SPECTRE_TEST_CASE("Unit.Elliptic.DG.Actions.InhomogeneousBoundaryConditions",
+    // - lower xi: | 0 3 6 ] -> eta
     // This expectation assumes the diagonal mass matrix approximation.
-    const DataVector source_expected{-6., 0., 0., -6., 0., 0., -18., 0., 12.};
+    const DataVector source_expected{0., 0., 0., 3., 0., 0., 18., 24., 36.};
```
Why are you changing the numbers?

nilsvu replied:
Because @nilsdeppe requested I make the solution depend on dimensions https://github.com/sxs-collaboration/spectre/pull/1652#discussion_r316264460

nilsvu commented:
@nilsdeppe could you check the fixups and complete your review please?

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/ConjugateGradient/InitializeElement.hpp:
```diff
@@ -34,55 +35,89 @@ struct InitializeResidual;
-    auto r = db::item_type<residual_tag>(get<operand_tag>(box));
+  template <typename DataBox, typename... InboxTags, typename ArrayIndex,
+            typename ActionList, typename ParallelComponent>
+  static auto apply(DataBox& box,
```
`db::DataBox`

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/ConjugateGradient/InitializeElement.hpp:
```diff
@@ -34,55 +35,89 @@ struct InitializeResidual;
+            typename Metavariables, typename ArrayIndex,
+            Requires<db::tag_is_retrievable_v<LinearSolver::Tags::HasConverged,
+                                              DataBox>> = nullptr>
+  static void apply(DataBox& box,
```
`db::DataBox`

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitorActions.hpp:
```diff
@@ -59,11 +59,14 @@ struct InitializeResidual {
                                  const gsl::not_null<double*>
                                      local_residual_square) noexcept {
           *local_residual_square = residual_square;
+          // Also setting this so re-initialization works.
```
this -> `ite...`

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/Gmres/InitializeElement.hpp:
```diff
@@ -47,49 +52,93 @@ struct InitializeElement {
-    const auto& q = get<operand_tag>(box);
+  template <typename DataBox, typename... InboxTags, typename ArrayIndex,
+            typename ActionList, typename ParallelComponent>
+  static auto apply(DataBox& box,
```
again

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/Gmres/InitializeElement.hpp:
```diff
@@ -47,49 +52,93 @@ struct InitializeElement {
+            db::item_type<orthogonalization_iteration_id_tag>{0},
+            std::move(basis_history),
+            db::item_type<LinearSolver::Tags::HasConverged>{}),
+        // Terminate algorithm for now. The reduction will be broadcast to the
```
I feel like expanding on `the reduction` would be good

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/Gmres/InitializeElement.hpp:
```diff
@@ -47,49 +52,93 @@ struct InitializeElement {
+            typename Metavariables, typename ArrayIndex,
+            Requires<db::tag_is_retrievable_v<LinearSolver::Tags::HasConverged,
+                                              DataBox>> = nullptr>
+  static void apply(DataBox& box,
```
again

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/Gmres/InitializeElement.hpp:
```diff
@@ -47,49 +52,93 @@ struct InitializeElement {
+            Requires<db::tag_is_retrievable_v<LinearSolver::Tags::HasConverged,
+                                              DataBox>> = nullptr>
+  static void apply(DataBox& box,
+                    const Parallel::ConstGlobalCache<Metavariables>& cache,
```
non const

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/Gmres/ResidualMonitorActions.hpp:
```diff
@@ -57,15 +57,31 @@ struct InitializeResidualMagnitude {
+                orthogonalization_history) noexcept {
           *local_residual_magnitude = *initial_residual_magnitude =
               residual_magnitude;
+          // Also setting this so re-initialization works.
```
this -> ite....

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ElementActions.hpp:
```diff
@@ -86,40 +87,43 @@ struct PerformStep {
+                     db::tag_is_retrievable_v<fields_tag, DataBox> and
+                     db::tag_is_retrievable_v<operand_tag, DataBox> and
+                     db::tag_is_retrievable_v<operator_tag, DataBox>> = nullptr>
   static auto apply(DataBox& box,
```
`db::DataBox`

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ElementActions.hpp:
```diff
@@ -136,35 +140,38 @@ struct UpdateFieldValues {
+                     db::tag_is_retrievable_v<residual_tag, DataBox>> = nullptr>
   static auto apply(DataBox& box,
-                    Parallel::ConstGlobalCache<Metavariables>& cache,
+                    const Parallel::ConstGlobalCache<Metavariables>& cache,
```
no

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitorActions.hpp:
```diff
@@ -34,31 +37,28 @@ struct UpdateOperand;
-  static auto apply(db::DataBox<DbTagsList>& box,
+ private:
+  using fields_tag = FieldsTag;
+  using residual_square_tag = db::add_tag_prefix<
```
why not keep these in the function?

nilsvu replied:
this makes it easier to do the `Requires` for the `apply` function

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitorActions.hpp:
```diff
@@ -34,31 +37,28 @@ struct UpdateOperand;
+            typename Metavariables, typename ArrayIndex,
+            Requires<db::tag_is_retrievable_v<residual_square_tag, DataBox>> =
+                nullptr>
+  static void apply(DataBox& box,
```
agadn

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitorActions.hpp:
```diff
@@ -92,61 +93,55 @@ struct InitializeResidual {
+            typename Metavariables, typename ArrayIndex,
+            Requires<db::tag_is_retrievable_v<residual_square_tag, DataBox>> =
+                nullptr>
+  static void apply(DataBox& box,
```
again

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitorActions.hpp:
```diff
@@ -92,61 +93,55 @@ struct InitializeResidual {
-  static auto apply(db::DataBox<DbTagsList>& box,
+ private:
+  using fields_tag = FieldsTag;
+  using residual_square_tag = db::add_tag_prefix<
```
same question

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/Gmres/ElementActions.hpp:
```diff
@@ -101,39 +102,43 @@ struct PerformStep {
-                                        DataBox>> = nullptr>
+ private:
+  using fields_tag = FieldsTag;
+  using operand_tag =
```
again

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/Gmres/ElementActions.hpp:
```diff
@@ -101,39 +102,43 @@ struct PerformStep {
+                     db::tag_is_retrievable_v<basis_history_tag, DataBox> and
+                     db::tag_is_retrievable_v<LinearSolver::Tags::IterationId,
+                                              DataBox>> = nullptr>
   static void apply(DataBox& box,
```
again

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/Gmres/ResidualMonitorActions.hpp:
```diff
@@ -99,35 +100,32 @@ struct InitializeResidualMagnitude {
+            typename Metavariables, typename ArrayIndex,
+            Requires<db::tag_is_retrievable_v<orthogonalization_history_tag,
+                                              DataBox>> = nullptr>
+  static void apply(DataBox& box,
```
please stop doing this. You can constrain the type, so constrain it.

nilsdeppe reviewed (changes requested):
I'm happy with the current fixups

nilsvu commented:
@nilsdeppe posted fixups for your comments. Please check if they resolve your issues. Thanks!

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/Gmres/ResidualMonitorActions.hpp:
```diff
@@ -79,7 +79,7 @@ struct InitializeResidualMagnitude {
           *local_residual_magnitude = *initial_residual_magnitude =
               residual_magnitude;
-          // Also setting this so re-initialization works.
+          // Also setting the following tags so re-initialization works.
```
I would suggest listing the tags explicitly so that if/when the code or comment gets moved it has a greater chance of providing context :)

nilsdeppe reviewed (commented):
I'm happy for you to squash everything.

nilsvu commented:
@nilsdeppe I had forgotten to drop the one commit we discussed above (https://github.com/sxs-collaboration/spectre/pull/1652#discussion_r316173172) so did that now.

@moxcodes @hrueter @nilsdeppe Travis is happy with this, so could you please take a final look and see if you are, too? Thanks for reviewing!
